File: src/core/manager.cpp
#include "core/manager.hpp"
#include <nlohmann/json.hpp>

namespace warden {
namespace core {

PasswordManager::PasswordManager(std::shared_ptr<Storage> storage)
    : storage_(std::move(storage)) {}

bool PasswordManager::createEntry(const std::string& service,
                                const std::string& username,
                                const std::string& password) {
    nlohmann::json entry = {
        {"service", service},
        {"username", username},
        {"password", password}
    };

    std::string json_str = entry.dump();
    std::vector<uint8_t> data(json_str.begin(), json_str.end());
    
    return storage_->write(service, data);
}

std::optional<PasswordEntry> PasswordManager::getEntry(const std::string& service) {
    auto data = storage_->read(service);
    if (!data) {
        return std::nullopt;
    }

    try {
        std::string json_str(data->begin(), data->end());
        auto entry = nlohmann::json::parse(json_str);

        return PasswordEntry{
            entry["service"].get<std::string>(),
            entry["username"].get<std::string>(),
            entry["password"].get<std::string>()
        };
    } catch (const std::exception&) {
        return std::nullopt;
    }
}

bool PasswordManager::updateEntry(const std::string& service,
                                const std::string& username,
                                const std::string& password) {
    nlohmann::json entry = {
        {"service", service},
        {"username", username},
        {"password", password}
    };

    std::string json_str = entry.dump();
    std::vector<uint8_t> data(json_str.begin(), json_str.end());
    
    return storage_->update(service, data);
}

bool PasswordManager::deleteEntry(const std::string& service) {
    return storage_->remove(service);
}

std::vector<std::string> PasswordManager::listEntries() {
    return storage_->listKeys();
}

} // namespace core
} // namespace warden

File: core/manager.py
from typing import Optional, Dict, List
from .storage import Storage

class PasswordManager:
    def __init__(self, storage: Storage):
        self.storage = storage

    def create_entry(self, service: str, username: str, password: str) -> bool:
        """Create a new password entry"""
        entry = {
            'service': service,
            'username': username,
            'password': password
        }
        return self.storage.write(service, entry)

    def get_entry(self, service: str) -> Optional[Dict]:
        """Retrieve an entry by service name"""
        return self.storage.read(service)

    def update_entry(self, service: str, username: str, password: str) -> bool:
        """Update an existing entry"""
        entry = {
            'service': service,
            'username': username,
            'password': password
        }
        return self.storage.update(service, entry)

    def delete_entry(self, service: str) -> bool:
        """Delete an entry"""
        return self.storage.delete(service)

    def list_entries(self) -> List[str]:
        """List all stored services"""
        return self.storage.list_services()
